### PR TITLE
incr.comp.: Some preparatory work for caching more query results.

### PR DESCRIPTION
**Pull request by michaelwoerister**
This PR
* adds and updates some encoding/decoding routines for various query result types so they can be cached later, and
* adds missing `[input]` annotations for a few `DepNode` variants.

The situation around having to explicitly mark dep-nodes/queries as inputs is not really satisfactory. I hope we can find a way of making this more fool-proof in the future.

r? @nikomatsakis

### CODE REVIEW

File: src/librustc_metadata/decoder.rs [hunk 1]
```diff
@@ -327,6 +328,14 @@ impl<'a, 'tcx> SpecializedDecoder<Span> for DecodeContext<'a, 'tcx> {
     }
 }
 
+impl<'a, 'tcx, T: Decodable> SpecializedDecoder<mir::ClearCrossCrate<T>>
+for DecodeContext<'a, 'tcx> {
+    #[inline]
+    fn specialized_decode(&mut self) -> Result<mir::ClearCrossCrate<T>, Self::Error> {
+        Ok(mir::ClearCrossCrate::Clear)
```
Comment (nikomatsakis) [inline src/librustc_metadata/decoder.rs]: that's...a cute trick

**Review by nikomatsakis — APPROVED** [review]

**Comment by nikomatsakis** [comment]
r=me after rebase

File: src/librustc_metadata/decoder.rs (hunk 1 above)
Comment (michaelwoerister) [reply src/librustc_metadata/decoder.rs]: Right? `:)`
I poached it from @arielb1.

**Comment by michaelwoerister** [comment]
Looks spurious.
@bors retry

**Comment by michaelwoerister** [comment]
@bors p=1

(A while back we decided that incr. comp. PRs get higher priority until the end of the year. Since there are 4 incr. comp. PRs open in parallel at the moment, with more to come, I'm taking advantage of this now)

**Pull request merged**